Map the I/O memory to the userspace

Tangzh33 opened the pull request:
We are currently working on enabling a lightweight desktop environment for Asterinas (issue #2008). A key step in this process is enabling the framebuffer for graphic output. The previous effort (PR #2017) has successfully allowed the kernel to utilize the framebuffer as one of the console outputs. This PR takes it a step further: implement the necessary `mmap` and `munmap` functions to support mapping the I/O memory to user space.

Thanks to PR #2042, we can start implementation directly from the system call `mmap`. Some design points may need further discussion.

1. Modification to the trait `FileIo` and the trait `FileLike`
    
    It is unavoidable to add support in the trait `FileIo` and the trait `FileLike`, to either expose more info about the file‘s `IoMem`, or steer the `mmap` process to file-specific functions. In this PR, I chose the former option, which is adding a function to expose the file’s  `IoMem`.
    
    From my perspective, the current solution is general enough to support most of the mmap operations of files with `IoMem`. If we choose the latter option, we must implement many duplicate operations like `vm_map_options.build()` for each specific file, instead of reusing existing code.
    
    Essentially, it is a trade-off between an increasingly longer `do_sys_mmap` function or duplicated implementations within each file’s `mmap`.
    
2. Ownership of `IoMem`
    
    After a discussion with @junyang-zh , this PR tracks the ownership of `IoMem` in `Vmar` and `VmSpace`. 
    
    - Track in `Vmar`: Keep a clone for possible future usage and align it with `MappedVmo`'s behaviors.
    - Track in `VmSpace`: The `ostd` should know the lifetime and reference count of `IoMem`.
3. Page fault and `mmap` populate
    
    This PR maps `IoMem`'s physical pages to the page table during `mmap`. Since the page table nodes for `IoMem` will ultimately be mapped to corresponding physical pages, completing the mapping in `mmap` is more efficient than mapping page by page in the page fault handler.
    

Finally, this is a draft PR and builds on previous efforts like PR #2042 and PR #1948. According to our community standards, it may have some flaws, and it is proposed to gather quick feedback.

vvvvsv commented on kernel/src/vm/vmar/mod.rs:
```diff
@@ -1053,6 +1084,17 @@ where
+        // elegant implementation.
+        if map_populate {
+            let mut rss_delta = RssDelta::new(&parent.0);
+            vm_mapping = vm_mapping.map_populate(parent.vm_space(), &mut rss_delta)?;
```
Do not use `?` to return an `Err` here. This will fix your
```
        // TODO: According to the mmap() manual of MAP_POPULATE, the mmap() call
        // doesn't fail if the mapping cannot be populated (for example, due to
        // limitations on the number of mapped huge pages when using MAP_HUGETLB).
```
in `vm_mappings.rs`.

vvvvsv commented on kernel/src/vm/vmar/mod.rs:
```diff
@@ -1042,8 +1067,14 @@ where
         // Build the mapping.
-        let vmo = vmo.map(|vmo| MappedVmo::new(vmo.to_dyn(), vmo_offset..vmo_limit));
-        let vm_mapping = VmMapping::new(
+        let vmo = if let Some(vmo) = vmo {
+            MappedVmObj::Vmo(MappedVmo::new(vmo.to_dyn(), vmo_offset..vmo_limit))
+        } else if let Some(iomem) = iomem {
+            MappedVmObj::IoMem(MappedIOMem::new(iomem, vmo_offset..vmo_limit))
```
We don't need to store a range in `MappedVmObj`. Instead, we can just store an `offset`, which equals to the `vmo_offset` here. See https://github.com/asterinas/asterinas/pull/2239.

vvvvsv commented on kernel/src/vm/vmar/vm_mapping.rs:
```diff
@@ -657,3 +777,86 @@ fn try_merge(left: &VmMapping, right: &VmMapping) -> Option<VmMapping> {
+
+    /// Returns the size of the mapped IO memory region in bytes.
+    #[expect(dead_code)]
+    pub(super) fn size(&self) -> usize {
+        self.range.len()
+    }
```
This method should return `self.io_mem.length()`.

vvvvsv replied:
I just found that I overlooked yesterday. This method should return `(self.offset..self.io_mem.length()).len()`. See
https://github.com/asterinas/asterinas/blob/a4c918efd737f5b7299b31e5a9646a03f8f17457/kernel/src/vm/vmar/vm_mapping.rs#L588-L595

Sorry for my oversight...

vvvvsv commented on kernel/src/vm/vmar/vm_mapping.rs:
```diff
@@ -630,20 +733,37 @@ fn try_merge(left: &VmMapping, right: &VmMapping) -> Option<VmMapping> {
+            MappedVmObj::Vmo(MappedVmo::new(l_vmo.vmo.dup().ok()?, range))
+        }
+        (MappedVmObj::IoMem(l_iomem), MappedVmObj::IoMem(r_iomem)) => {
+            let is_offset_contiguous = r_iomem.range.start - l_iomem.range.start == left.map_size()
```
Do not use `r_iomem.range.start - l_iomem.range.start` here, since it may cause a underflow panic. `l_iomem.range.start + left.map_size() == r_iomem.range.start` would be ok.

vvvvsv commented on kernel/src/vm/vmar/vm_mapping.rs:
```diff
@@ -630,20 +733,37 @@ fn try_merge(left: &VmMapping, right: &VmMapping) -> Option<VmMapping> {
+                return None;
+            }
+            let range = l_iomem.range.start..l_iomem.range.end.max(r_iomem.range.end);
+            // FIXME: Invite @vvvsv to check this.
+            MappedVmObj::IoMem(MappedIOMem::new(l_iomem.io_mem.clone(), range))
```
It looks good to me.

vvvvsv commented on kernel/src/vm/vmar/vm_mapping.rs:
```diff
@@ -153,6 +159,85 @@ impl VmMapping {
+                    continue;
+                }
+
+                // Prepare the page using the same logic as page fault handler
+                let (frame, is_readonly) = match self.prepare_page(page_addr, false) {
```
I don't think we should simply prepare a read-only page using `self.prepare_page(page_addr, write: false)`. In this way, the `mmap`s below can't be properly handled.

```c
mmap(NULL, len, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_POPULATE, -1, 0);
int fd = open("file", O_RDWR);
mmap(NULL, len, PROT_READ | PROT_WRITE, MAP_SHARED | MAP_POPULATE, fd, 0);
mmap(NULL, len, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_POPULATE, fd, 0);
```

Tangzh33 commented on ostd/src/mm/vm_space.rs:
```diff
@@ -416,8 +533,103 @@ pub(super) fn get_activated_vm_space() -> *const VmSpace {
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum MappedItem {
```
The reason why it is `pub(crate)` is the same as `struct VmItem`.

> Another issue is that UserPtConfig is pub(crate) but it has type Item = VmItem in its implementation, and if VmItem is pub(super), it will cause a visibility problem.

lrh2000 commented on ostd/src/mm/page_table/test.rs:
```diff
@@ -252,10 +253,15 @@ mod page_properties {
         let queried = page_table.page_walk(virtual_range.start + 100).unwrap().1;
-        assert_eq!(queried, prop);
+
+        // When using VmItem::new_tracked(), it's always a tracked frame, not I/O memory
```
```suggestion
        // When using `VmItem::new_tracked`, it's always a tracked frame, not I/O memory
```

lrh2000 commented on ostd/src/mm/vm_space.rs:
```diff
@@ -292,29 +307,91 @@ impl<'a> CursorMut<'a> {
+        let mut current_paddr = (io_mem.paddr() + offset).align_down(PAGE_SIZE);
+        let paddr_end = min(
+            io_mem.paddr() + io_mem.length().align_up(PAGE_SIZE),
+            (current_paddr + len).align_up(PAGE_SIZE),
```
The alignment is already checked. No need to align here.

lrh2000 commented on ostd/src/mm/vm_space.rs:
```diff
@@ -292,29 +307,91 @@ impl<'a> CursorMut<'a> {
+            io_mem.paddr() + io_mem.length().align_up(PAGE_SIZE),
+            (current_paddr + len).align_up(PAGE_SIZE),
+        );
+        while current_paddr < paddr_end {
```
Use `for` loop can be clearer than a `while` loop.

lrh2000 reviewed (commented):
I don't believe the current `MappedVmObject` design is ideal. See below for details.

Besides, the commits do not seem to be atomic and can indeed create unnecessary confusion if reviewing by commit.

Tangzh33 commented on kernel/src/vm/vmar/mod.rs:
```diff
@@ -983,6 +996,33 @@ impl<R1> VmarMapOptions<'_, R1, Rights> {
+                self.inode = Some(inode);
+            }
+            MemoryToMap::IoMem(io_mem) => {
+                assert!(len <= io_mem.length().align_up(PAGE_SIZE));
```
Sure, I have tested this on Linux, and you are correct. For files that are backed by I/O memory, the `mmap` function will not fail if the sum of `offset` and `size` exceeds the actual file size. However, if the program attempts to access a region beyond the limit of the I/O memory size, it will be killed due to a segmentation fault.

lrh2000 commented on kernel/src/vm/vmar/mod.rs:
```diff
@@ -966,6 +969,12 @@ impl<'a, R1, R2> VmarMapOptions<'a, R1, R2> {
     }
+
+    /// Sets whether the mapping should be populated immediately.
+    pub fn map_populate(mut self, populate: bool) -> Self {
```
Some of the code in the last commit is unreachable without properly handling the `MAP_POPULATE`. Adding support for `MAP_POPULATE` requires additional work, such as adding and enabling regression tests. It is not generally related to this PR at all.

I think in this PR, we can have add a `populate_device(&self, io_mem: IoMem)` method. We can extend that method to a general `populate()` method in a future PR.

```rust
fn populate_device(&self, io_mem: IoMem) {
    debug_assert!(matches!(self.mapped_mem, MappedMemory::Device));

    let cursor = ...;
    cursor.map_iomem(...);
}
```

Tangzh33 replied:
Thank you for your suggestions. I have removed the untouched logic. This PR is now ready for review again; please drop your comments at your convenience:-)

lrh2000 reviewed (commented):
Thanks for the revision! The code looks much better to me now. There are a few more comments, which I hope are still minor, that I didn't notice in previous rounds.

Tangzh33 commented on kernel/src/fs/file_handle.rs:
```diff
@@ -49,6 +51,16 @@ pub trait FileLike: Pollable + Send + Sync + Any {
+    /// returns a [`MemoryToMap`] that describes the memory object to be mapped
+    /// at the specified offset and length. The caller is responsible for
+    /// creating the actual memory mapping using this information.
+    fn mmap(&self, len: usize, offset: usize) -> Result<MemoryToMap> {
```
Sure, the `len` and the `offset` are written by intuition, and the GUI code we're supporting doesn't touch their corresponding logic. Although your proposal for their possible use seems reasonable, there are no example codes to test it at this point. I will remove these variables and let the following PR handle them properly if there is actual demand.

lrh2000 commented on kernel/src/vm/vmar/mod.rs:
```diff
@@ -1010,6 +1058,21 @@ where
+        //     (None, Some(vmo)) => (Some(vmo.to_dyn()), None, None),
+        //     (None, None) => (None, None, None), // Anonymous mapping
+        //     _ => panic!("Invalid combination of memory_to_map and vmo"),
+        // };
```
What's this?

lrh2000 commented on kernel/src/vm/vmar/mod.rs:
```diff
@@ -1147,22 +1147,18 @@ where
-        // FIXME: Add a function to populate the memory in vm_mapping.
+        // Populate device memory if needed before adding to VMAR
+        //
+        // We have to map before inserting the VmMapping into the tree, otherwise
```
```suggestion
        // We have to map before inserting the `VmMapping` into the tree, otherwise
```

lrh2000 commented on kernel/src/vm/vmar/vm_mapping.rs:
```diff
@@ -163,6 +164,35 @@ impl VmMapping {
+    ) -> Result<()> {
+        debug_assert!(matches!(self.mapped_mem, MappedMemory::Device));
+        if !matches!(self.mapped_mem, MappedMemory::Device) {
+            panic!("populate_device called on non-device mapping");
```
No need to do this. This is an internal method so I think `debug_assert!` is enough.

lrh2000 commented on ostd/src/mm/vm_space.rs:
```diff
@@ -396,16 +384,11 @@ impl<'a> CursorMut<'a> {
+                        // Flush the TLB entry for the current address, but DO
+                        // NOT drop the corresponding `IoMem`. This is because
+                        // we manage the range of I/O as a whole, but the
+                        // frames handled here might be one segment of it.
```
Be more careful in rebasing the changes? I think you've updated the comments but you restore the comments to the old version...

lrh2000 commented on kernel/src/vm/vmar/vm_mapping.rs:
```diff
@@ -314,8 +359,20 @@ impl VmMapping {
+                return Ok((FrameAllocOptions::new().alloc_frame()?.into(), is_readonly));
+            }
+            MappedMemory::Device => {
+                // Device memory should not go through the page fault handler
+                return Err(VmoCommitError::Err(Error::with_message(
+                    Errno::EFAULT,
+                    "Device memory mapping should not go through the page fault handler",
+                )));
```
The wording is inaccurate:
```suggestion
                // Device memory is populated when the memory mapping is created.
                return Err(VmoCommitError::Err(Error::with_message(
                    Errno::EFAULT,
                    "device memory page faults cannot be resolved",
                )));
```

lrh2000 commented on kernel/src/vm/vmar/vm_mapping.rs:
```diff
@@ -265,6 +308,11 @@ impl VmMapping {
+                Some(VmQueriedItem::MappedIoMem { .. }) => {
+                    // The page of I/O memory should not go through the
+                    // page fault handler.
+                    return_errno_with_message!(Errno::EFAULT, "I/O memory page fault");
```
Same as above.

lrh2000 reviewed (commented):
Thanks for your efforts in addressing the comments again.

I noticed some rebasing errors. To make things easier for you and the reviewer, I suggest squashing the last three commits and splitting them again according to whether the changes are in OSTD or the kernel.
<img width="809" height="342" alt="image" src="https://github.com/user-attachments/assets/72bcfcb6-315f-49e6-a067-1ca8ba2bd0e6" />

These three commits will be easier and better to review if they look like this:
 - Add OSTD support for mapping `IoMem` in userspace
 - Add kernel support for mapping `IoMem` via files

lrh2000 reviewed (commented):
Thanks for addressing the comments in a timely manner!

I should really notice (TBD) a bit earlier. Everything else look quite good to me.

lrh2000 commented on kernel/src/vm/vmar/mod.rs:
We cannot expand device mappings at `mremap`. Linux rejects this scenario here: https://elixir.bootlin.com/linux/v6.16/source/mm/mremap.c#L1325

This means something like:
```patch
diff --git a/kernel/src/vm/vmar/mod.rs b/kernel/src/vm/vmar/mod.rs
index 1dcffea1b..93a6bbdf7 100644
--- a/kernel/src/vm/vmar/mod.rs
+++ b/kernel/src/vm/vmar/mod.rs
@@ -600,11 +600,14 @@ impl Vmar_ {
         let mut inner = self.inner.write();
         let mut rss_delta = RssDelta::new(self);
 
-        if inner.vm_mappings.find_one(&old_addr).is_none() {
+        let Some(old_mapping) = inner.vm_mappings.find_one(&old_addr) else {
             return_errno_with_message!(
                 Errno::EFAULT,
                 "remap: there is no mapping at the old address"
             )
+        };
+        if new_size > old_size && !old_mapping.can_expand() {
+            return_errno_with_message!(Errno::EFAULT, "remap: device mappings cannot be expanded");
         }
 
         // Shrink the old mapping first.
```

Tangzh33 commented on ostd/src/mm/vm_space.rs:
```diff
@@ -201,6 +201,33 @@ impl Default for VmSpace {
+    /// This method searches through the [`VmSpace`]'s iomems set to find
+    /// an [`IoMem`] object that contains the specified physical address.
+    /// Returns the [`IoMem`] and the offset within that [`IoMem`].
+    pub fn find_iomem_for_paddr(&self, paddr: Paddr) -> Option<(IoMem, usize)> {
```
Thanks for your helpful analysis and careful review of the code document's wording. I finally chose the `find_iomem_by_paddr` version because I prefer not to modify `VmQueriedItem`, as the `IoMem` field remains unused in most of our cases.

I have refined the documents and addressed the `pub` issues based on your comments; please check if there are any remaining problems.

lrh2000 commented on kernel/src/vm/vmar/mod.rs:
```diff
@@ -1384,4 +1385,143 @@ mod test {
+
+    /// Helper function to create a dummy `IoMem` for testing.
+    #[expect(unsafe_code)]
+    fn create_dummy_iomem(paddr: usize, size: usize) -> IoMem {
```
Remove this. This isn't safe at all. Use `IoMem::acquire` instead. You may need to choose a larger address to ensure that it succeeds (it should be larger than the maximum physical address).

lrh2000 commented on kernel/src/vm/vmar/mod.rs:
```diff
@@ -1384,4 +1385,143 @@ mod test {
+        let page_property = PageProperty::new_user(PageFlags::RW, CachePolicy::Uncacheable);
+        let preempt_guard = disable_preempt();
+
+        // Create and map IoMem instead of a frame.
```
Be consistent to existing code:
```suggestion
        // Creates and maps an `IoMem` instead of a frame.
```

lrh2000 commented on kernel/src/vm/vmar/mod.rs:
```diff
@@ -1384,4 +1385,143 @@ mod test {
+            let mut child_cursor = child_space.cursor_mut(&preempt_guard, &cow_range).unwrap();
+            let mut parent_cursor = vm_space.cursor_mut(&preempt_guard, &cow_range).unwrap();
+            let num_copied = cow_copy_pt(&mut parent_cursor, &mut child_cursor, cow_range.len());
+            assert_eq!(num_copied, 0); // IoMem pages are not "copied" in the same sense as RAM
```
```suggestion
            assert_eq!(num_copied, 0); // `IoMem` pages are not "copied" in the same sense as RAM pages.
```

lrh2000 commented on kernel/src/vm/vmar/mod.rs:
```diff
@@ -1384,4 +1385,143 @@ mod test {
+            .unwrap()
+            .unmap(map_range.len());
+
+        // Confirms that the child VA remains mapped (IoMem mappings should persist).
```
No child pages should be affected by unmapping pages in the parent page table, which is not related to whether the pages are `IoMem` pages or not.
```suggestion
        // Confirms that the child VA remains mapped.
```

lrh2000 commented on kernel/src/vm/vmar/mod.rs:
```diff
@@ -1384,4 +1385,143 @@ mod test {
+            .unwrap()
+            .map_iomem(iomem_clone_for_assert, page_property, PAGE_SIZE, 0);
+
+        // Confirms that the sibling mapping points back to the original IoMem's physical address.
```
```suggestion
        // Confirms that the sibling mapping points back to the original `IoMem`'s physical address.
```

lrh2000 commented on ostd/src/io/io_mem/mod.rs:
```diff
@@ -76,7 +76,7 @@ impl IoMem {
     /// - Reading from or writing to I/O memory regions may have side effects. Those side effects
     ///   must not cause soundness problems (e.g., they must not corrupt the kernel memory).
-    pub(crate) unsafe fn new(range: Range<Paddr>, flags: PageFlags, cache: CachePolicy) -> Self {
+    pub unsafe fn new(range: Range<Paddr>, flags: PageFlags, cache: CachePolicy) -> Self {
```
Don't mark unsafe functions as public. Revert this change.

lrh2000 commented on ostd/src/mm/test.rs:
```diff
@@ -499,6 +500,12 @@ mod vmspace {
     }
 
+    /// Helper function to create a dummy `IoMem` for testing.
+    fn create_dummy_iomem(paddr: Paddr, size: usize) -> IoMem {
```
Similar to above. Please use `IoMem::acquire`.

lrh2000 reviewed (commented):
Thanks for the updates. Most comments are just revising the comments. I hope these can be my final comments.

junyang-zh commented on ostd/src/mm/vm_space.rs:
```diff
@@ -421,8 +527,83 @@ pub(super) fn get_activated_vm_space() -> *const VmSpace {
+    }
+
+    /// Converts the `VmItem` into a `VmQueriedItem`.
+    fn into_queried_item(self) -> VmQueriedItem {
+        match self.mapped_item {
+            MappedItem::TrackedFrame(frame) => VmQueriedItem::MappedRam {
+                frame,
+                prop: self.prop,
+            },
+            MappedItem::UntrackedIoMem { paddr, level } => {
+                debug_assert_eq!(level, 1);
+                VmQueriedItem::MappedIoMem {
+                    paddr,
+                    prop: self.prop,
+                }
+            }
+        }
+    }
+}
+
+impl From<VmItem> for VmQueriedItem {
+    fn from(item: VmItem) -> Self {
+        item.into_queried_item()
+    }
+}
```
Perhaps by moving all the implementation from `into_queried_item` to `impl From<VmItem> for VmQueriedItem` and removing `into_queried_item`, the code would be more straightforward and concise. I don't see any usage of `into_queried_item` other than here.

Tangzh33 commented:
@boterinas rerun

junyang-zh reviewed (approved):
LGTM! Thanks both @Tangzh33 @lrh2000 for your rigorous work!

tatetian commented:
I have some last minute comments. Please hold on before merging this PR.

tatetian commented on kernel/src/fs/file_handle.rs:
```diff
@@ -49,6 +51,16 @@ pub trait FileLike: Pollable + Send + Sync + Any {
+    /// returns a [`MemoryToMap`] that describes the memory object to be mapped.
+    /// The caller is responsible for creating the actual memory mapping using
+    /// this information.
+    fn mmap(&self) -> Result<MemoryToMap> {
```
I think that the name of this method could be confusing to many people because the signature and the semantics of this `mmap` method is quite different from that of the `mmap` system call. This method does not create any mapping at all. We can't call a method `mmap` simply because it is primarily used for `mmap`.

The name `MemoryToMap` is also inappropriate to me. You can't `Inode` is a kind of "memory" to map. It is not memory.

My suggestion is to rename `MemoryToMap` to `Mappable` and this method to `mappable`.

```rust
pub trait FileLike: Pollable + Send + Sync + Any {
    /// Obtains the mappable object to map this file into the user address space.
    ///
    /// If this file has a corresponding mappable object of [`Mappable`],
    /// then it can be either an inode or a MMIO region.
    fn mappable(&self) -> Result<Mappable> { ... }
}

/// An object that may be memory mapped into the user address space.
pub enum Mappable {
    /// An inode object.
    Inode(Arc<dyn Inode>),
    /// An MMIO region.
    #[expect(dead_code)]
    IoMem(IoMem),    
}
```

tatetian commented on kernel/src/fs/file_handle.rs:
```diff
@@ -149,3 +157,16 @@ impl dyn FileLike {
+    /// The mapping maps to the page cache.
+    PageCache(Arc<dyn Inode>),
+    /// The mapping maps to the I/O memory.
+    #[expect(dead_code)]
```
Why is this `dead_code`? This variant is used.

tatetian reviewed (commented):
The code looks most great. I appreciate the effort that the author and the reviewers put into this PR.

Tangzh33 commented:
@tatetian @lrh2000 @junyang-zh @vvvvsv Thank you all for your rigorous review and the patience you showed during these iterations. I've learned a lot from you guys, not only about the coding techniques, but also about the professional attitude. I have addressed all the remaining comments. Please take a look to see if there are any extra problems.

lrh2000 commented on kernel/src/fs/file_handle.rs:
```diff
@@ -49,6 +51,14 @@ pub trait FileLike: Pollable + Send + Sync + Any {
+    /// Obtains the mappable object to map this file into the user address space.
+    ///
+    /// If this file has a corresponding mappable object of [`Mappable`],
+    /// then it can be either an inode or a MMIO region.
```
```suggestion
    /// then it can be either an inode or an MMIO region.
```

lrh2000 commented on kernel/src/fs/inode_handle/mod.rs:
```diff
@@ -167,6 +167,21 @@ impl InodeHandle_ {
+            // return the file-specific mappable object.
+            file_io.mappable()
+        } else {
+            return_errno_with_message!(Errno::EINVAL, "The file is not mappable");
```
- Lowercase the first letter in the error message:
```suggestion
            return_errno_with_message!(Errno::EINVAL, "the file is not mappable");
```

- I think since you change this, you should change the error messages in the default implementation of `FileIo::mapped` and `FileLike::mapped` as well. Any reason not to do that?

lrh2000 commented on kernel/src/vm/vmar/mod.rs:
```diff
@@ -1260,7 +1328,141 @@ mod test {
+    #[ktest]
+    fn test_cow_copy_pt_iomem() {
+        /// A very large address (1TiB) beyond typical physical memory for testing.
+        const IOMEM_PADDR: usize = 0x100_000_000_000;
```
Please insert a blank line below this constant.

tatetian reviewed (approved):
LGTM. Thanks for the contribution!
